### ISSUE TITLE
How to import an existing AST into ppxlib?

### ISSUE THREAD

**Issue by emillon** [ocaml-ppx/ppxlib#185]
Hi,

I have a piece of code that builds a 4.08 AST using OMP types (it's a patched version of the OCaml parser from 4.08). I would like to convert it to a ppxlib AST.

At the moment I'm using:

    module Conv = Ppxlib_ast.Select_ast (Ppxlib_ast__.Versions.OCaml_408)
    module To_ppxlib = Conv.Of_ocaml

But I can't find OCaml_408 in the public API (in ppxlib, only `Selected_ast` is exposed, and in OMP2, only the migrations are there).

Thanks!

**Comment by emillon** [comment]
Is adding `module Versions = Versions` to `ppxlib_ast.ml` an acceptable fix, or should a tighter interface be used?

**Comment by pitag-ha** [comment]
This reply comes late for the ocamlformat port to ppxlib, but as a general reply: if `Versions` is helpful also for others than ppxlib, I could move that logic into Astlib (before upstreaming Astlib into the compiler). Do you (or anyone else who reads this) know of any other project that would benefit from that?

**Comment by emillon** [comment]
I think it would only benefit projects that use a fixed parser (rather than the one from compiler-libs). This is getting fixed in ocamlformat (by not using ppxlib in the area that uses a fixed parser). I think there was a similar situation with bucklescript at some point?

**Comment by pitag-ha** [comment]
Oh right, for the bucklescript support we were in a similar situation. There, we only needed access to general conversions from within the ppxlib driver though, so I only added them to `Ppxlib_ast`, which is (meant to be) internal.

I think I'd leave it as is for now, but I keep this in mind in case it comes up again.

**Comment by pitag-ha** [comment]
Just to add a side-note about what this issue is about (I've already seen a couple of people who thought this situation applied to their projects):

The issue described here only applies to you if you need to handle a fixed AST version, which is not ppxlib's AST version. That's very rare. If you only need to deal with the ppxlib AST, you can do the conversions via the `Ppxlib_ast.Selected_ast` module, using one of its submodules [`Of_ocaml`](https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib_ast__/Import/Selected_ast/Of_ocaml/index.html) or [`To_ocaml`](https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib_ast__/Import/Selected_ast/To_ocaml/index.html). E.g. if you need to convert a `structure` of the compiler's AST to a `structure` of ppxlib's AST, you can use
```
Ppxlib_ast.Selected_ast.Of_ocaml.copy_structure
```

**Comment by emillon** [comment]
Good point! I renamed the issue to make it clearer. Also, since ocamlformat does not require this anymore I'm fine with closing it.
